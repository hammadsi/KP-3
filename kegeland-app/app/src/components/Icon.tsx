import React from 'react';
import {StyleProp, StyleSheet, TextStyle} from 'react-native';
import {useTheme} from 'react-native-paper';
import MaterialIcon from 'react-native-vector-icons/MaterialCommunityIcons';

export type IconProps = {
  color?: string;
  icon: string;
  size?: number;
  style?: StyleProp<TextStyle>;
};

/**
 * Component for rendering a icon.
 * @param props the props
 * @see {@link IconProps}
 * @see {@link MaterialIcon}
 */
const Icon: React.FC<IconProps> = (props) => {
  const {colors} = useTheme();
  return (
    <MaterialIcon
      color={props.color || colors.muted}
      size={props.size || 20}
      name={props.icon}
      style={[styles.icon, props.style]}
    />
  );
};

const styles = StyleSheet.create({
  icon: {alignSelf: 'center'},
});

export default Icon;
