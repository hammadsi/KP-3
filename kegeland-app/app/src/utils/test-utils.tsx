/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
