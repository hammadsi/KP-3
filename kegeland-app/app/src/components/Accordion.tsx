import React, {Children} from 'react';
import {StyleProp, StyleSheet, TextStyle, View} from 'react-native';
import {Divider, List, useTheme} from 'react-native-paper';

import Icon from './Icon';

export type AccordionProps = Omit<
  React.ComponentProps<typeof List.Accordion>,
  'expanded' | 'onPress' | 'left'
> & {
  icon?: string | React.ReactNode;
  iconSize?: number;
  iconStyle?: StyleProp<TextStyle>;
};

/**
 * Component for rendering an accordion menu.
 * This component has inbuilt functionality for hiding/displaying it's sub-components.
 * @param props the props
 * @see {@link AccordionProps}
 * @see {@link List.Accordion}
 */
const Accordion: React.FC<AccordionProps> = (props) => {
  const {colors} = useTheme();
  const [expanded, setExpanded] = React.useState(true);

  const handlePress = () => setExpanded(!expanded);

  /**
   * Render the an icon to the left if enabled
   */
  const renderLeft = () => {
    if (props.icon) {
      if (React.isValidElement(props.icon)) {
        return (
          <View
            style={[
              {width: props.iconSize, height: props.iconSize},
              styles.icon,
            ]}>
            {props.icon}
          </View>
        );
      } else {
        return (
          <Icon
            color={colors.primary}
            icon={props.icon as string}
            size={props.iconSize}
            style={props.iconStyle}
          />
        );
      }
    }
    return undefined;
  };

  return (
    <List.Accordion
      {...props}
      left={renderLeft}
      style={[
        {
          backgroundColor: colors.elevation.elevation1,
        },
        styles.accordion,
        props.style,
      ]}
      titleStyle={[{color: colors.text}, props.titleStyle]}
      expanded={expanded}
      onPress={handlePress}>
      {expanded && Children.toArray(props.children).length > 0 && <Divider />}
      <View style={[styles.content]}>{props.children}</View>
    </List.Accordion>
  );
};

Accordion.defaultProps = {
  iconSize: 24,
};

const styles = StyleSheet.create({
  accordion: {
    paddingHorizontal: 0,
    paddingLeft: 8,
  },
  icon: {
    justifyContent: 'center',
  },
  content: {
    paddingLeft: 12,
  },
});

export default Accordion;
