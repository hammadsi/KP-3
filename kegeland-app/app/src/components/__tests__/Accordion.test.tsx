import React from 'react';
import {Text} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import {render} from '~utils/test-utils';

import Accordion from '../Accordion';

describe('Test Accordion-component', () => {
  it('should render correctly', () => {
    const component = (
      <Accordion title="Test accordion">
        <Text>Test</Text>
      </Accordion>
    );
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render with icon given as string', () => {
    const component = (
      <Accordion title="Test accordion" icon="account-circle-outline">
        <Text>Test</Text>
      </Accordion>
    );
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render with icon given as a component', () => {
    const component = (
      <Accordion
        title="Test accordion"
        icon={<Icon size={30} color="#0081cb" name="chevron-left" />}>
        <Text>Test</Text>
      </Accordion>
    );
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
