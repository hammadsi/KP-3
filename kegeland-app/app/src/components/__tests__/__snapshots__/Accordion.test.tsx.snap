// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Accordion-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#f6f6f6",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
            "expanded": true,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "padding": 8,
              },
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "paddingHorizontal": 0,
                  "paddingLeft": 8,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "margin": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "#00b0ff",
                    },
                    Array [
                      Object {
                        "color": "#000000",
                      },
                      undefined,
                    ],
                  ],
                ]
              }
            >
              Test accordion
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "margin": 8,
                },
                undefined,
              ]
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅃
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "height": 0.5,
          },
          undefined,
          Array [
            Object {
              "paddingLeft": 64,
            },
            undefined,
          ],
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "paddingLeft": 64,
          },
          Array [
            Object {
              "paddingLeft": 12,
            },
          ],
        ]
      }
    >
      <Text>
        Test
      </Text>
    </View>
  </View>
</View>
`;

exports[`Test Accordion-component should render with icon given as a component 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#f6f6f6",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
            "expanded": true,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "padding": 8,
              },
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "paddingHorizontal": 0,
                  "paddingLeft": 8,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "height": 24,
                  "width": 24,
                },
                Object {
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#0081cb",
                    "fontSize": 30,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅁
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "margin": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "#00b0ff",
                    },
                    Array [
                      Object {
                        "color": "#000000",
                      },
                      undefined,
                    ],
                  ],
                ]
              }
            >
              Test accordion
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "margin": 8,
                },
                undefined,
              ]
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅃
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "height": 0.5,
          },
          undefined,
          Array [
            Object {
              "paddingLeft": 64,
            },
            undefined,
          ],
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "paddingLeft": 64,
          },
          Array [
            Object {
              "paddingLeft": 12,
            },
          ],
        ]
      }
    >
      <Text>
        Test
      </Text>
    </View>
  </View>
</View>
`;

exports[`Test Accordion-component should render with icon given as string 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#f6f6f6",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
            "expanded": true,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "padding": 8,
              },
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "paddingHorizontal": 0,
                  "paddingLeft": 8,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#00b0ff",
                  "fontSize": 24,
                },
                Array [
                  Object {
                    "alignSelf": "center",
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            󰭕
          </Text>
          <View
            style={
              Array [
                Object {
                  "margin": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "#00b0ff",
                    },
                    Array [
                      Object {
                        "color": "#000000",
                      },
                      undefined,
                    ],
                  ],
                ]
              }
            >
              Test accordion
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "margin": 8,
                },
                undefined,
              ]
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅃
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "height": 0.5,
          },
          undefined,
          Array [
            Object {
              "paddingLeft": 64,
            },
            undefined,
          ],
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "paddingLeft": 64,
          },
          Array [
            Object {
              "paddingLeft": 12,
            },
          ],
        ]
      }
    >
      <Text>
        Test
      </Text>
    </View>
  </View>
</View>
`;
